core/banking/gov_transfers.go

```diff
@@ -0,0 +1,271 @@
+package banking
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"code.vegaprotocol.io/vega/core/events"
+	"code.vegaprotocol.io/vega/core/types"
+	"code.vegaprotocol.io/vega/libs/num"
+	"code.vegaprotocol.io/vega/logging"
+	vegapb "code.vegaprotocol.io/vega/protos/vega"
+)
+
+func (e *Engine) distributeScheduledGovernanceTransfers(ctx context.Context) {
+	timepoints := []int64{}
+	now := e.timeService.GetTimeNow()
+	for k := range e.scheduledGovernanceTransfers {
+		if now.UnixNano() >= k {
+			timepoints = append(timepoints, k)
+		}
+	}
+
+	for _, t := range timepoints {
+		transfers := e.scheduledGovernanceTransfers[t]
+		for _, gTransfer := range transfers {
+			amt, err := e.processGovernanceTransfer(ctx, gTransfer)
+			if err != nil {
+				gTransfer.Status = types.TransferStatusStopped
+				e.broker.Send(events.NewGovTransferFundsEventWithReason(ctx, gTransfer, amt, err.Error()))
+			}
+		}
+		delete(e.scheduledGovernanceTransfers, t)
+	}
+}
+
+func (e *Engine) distributeRecurringGovernanceTransfers(ctx context.Context, newEpoch uint64) {
+	var (
+		transfersDone = []events.Event{}
+		doneIDs       = []string{}
+	)
+
+	for _, gTransfer := range e.recurringGovernanceTransfers {
+		if gTransfer.Config.RecurringTransferConfig.StartEpoch > newEpoch {
+			continue
+		}
+
+		amount, err := e.processGovernanceTransfer(ctx, gTransfer)
+		if err != nil {
+			e.log.Error("error calculating transfer amount for governance transfer", logging.Error(err))
+			gTransfer.Status = types.TransferStatusStopped
+			transfersDone = append(transfersDone, events.NewGovTransferFundsEventWithReason(ctx, gTransfer, amount, err.Error()))
+			doneIDs = append(doneIDs, gTransfer.ID)
+			continue
+		}
+
+		if gTransfer.Config.RecurringTransferConfig.EndEpoch != nil && *gTransfer.Config.RecurringTransferConfig.EndEpoch == e.currentEpoch {
```

Comment:
Minor: we check against `e.currentEpoch` here, but in the `StartEpoch` check against `newEpoch`. `newEpoch` is `e.currentEpoch` because thats what we pass in, so maybe bin `newEpoch`?